Custom repo templates

jasonkylefrank opened the pull request:
### Purpose of the pull request
This PR adds the ability to specify a _custom repository_ as the starting point (i.e., template) for the created Next.js app, instead of _only_ allowing the Next.js examples for the starting point.  I wanted to be able to _create_ other templates, push them to GitHub, and then use those as my starting-point template for `create-next-app`.  I did not necessarily want to get them added to the example set in Next.js's repo.  This PR solves that issue.

### Testing this code
Here is how you can test it without installing it.   Notice the new **--clone-repo** option:

$ `node cli --clone-repo https://github.com/<username>/<reponame>  output-dir`

This will check to see if the supplied URL is a git repo.  If it is, it proceeds to with a similar routine that happens when an "example" is specified:  It creates the output directory, and then it clones the repository into the new directory, followed by the other normal activities, such as running `npm install`, etc. 

Obviously if this package was installed it should be called like this:

$ `create-next-app --clone-repo https://github.com/<username>/<reponame>  output-dir`

stephenmathieson commented on cli.js:
```diff
@@ -18,13 +18,16 @@ program
     projectName = name
   })
   .option('-e, --example <example-path>', messages.exampleHelp())
+  .option('-c, --clone-repo <clone-repo-path>', messages.cloneRepoHelp())
```
I'm not sure "clone-repo" is self-explanatory enough for this. Perhaps a `--template` option would make more sense?

Open to discussion here.

jasonkylefrank replied:
@stephenmathieson 
I'm open to other terms.   I agree that including "template" might be helpful.  The reason I thought to at least include the word "clone" somewhere in the name is to communicate that this would in-fact do a regular git clone on the repo - which obviously means they will get all branches and history, etc.  

So maybe we could use a name that included both "template" and "clone"?   

Maybe something like:
 * `--clone-template`
 * `--clone-template-repo`
 * ?

Again, I'm open to other ideas for the name.  😀

stephenmathieson commented on lib/index.js:
```diff
@@ -25,6 +31,11 @@ module.exports = function createNextApp(opts) {
+    cloneCustomRepo({
+      projectName: projectName,
+      repoPath: opts.cloneRepo
+    }).then(installWithMessageFactory(opts))
```
I know errors are not being properly handled elsewhere, but it would be great if you could add a `.catch()` block here to prevent the process from crashing unexpectedly.

Not a blocker tho, since this isn't being done in the LOCs above and below.

jasonkylefrank replied:
@stephenmathieson 
I'm holding off on this one for now since I'm blocked on the promise issue (see the [other comment thread](https://github.com/segmentio/create-next-app/pull/48#discussion_r223020458)).

stephenmathieson commented on lib/utils/clone-repo.js:
```diff
@@ -0,0 +1,52 @@
+
+// Ensure the given repo exists
+const validateRepoName = repoPath =>
+  new Promise((resolve, reject) => {
```
This outer `Promise` is unnecessary. Instead, we can do something like:

```js
const validateRepoName = repoPath => git().raw(...).catch(err => {
  output.error(err)
  throw err
})
```

jasonkylefrank replied:
@stephenmathieson 
Fixed now.  It will come with a batch of updates when I update this PR.

For others trying to understand this situation, it turns out that the `git().raw()` method uses a promise to do its work already (see [this](https://github.com/steveukx/git-js/blob/ef368c87fb4e1be5bbd40b1644cc8475e7bd6915/src/git.js#L1385) and [this](https://github.com/steveukx/git-js/blob/master/src/util/deferred.js)).

jasonkylefrank replied:
@stephenmathieson Actually this is still an issue - see the [other comment thread](https://github.com/segmentio/create-next-app/pull/48#discussion_r223020458).

stephenmathieson commented on lib/utils/clone-repo.js:
```diff
@@ -0,0 +1,52 @@
+  })
+
+const clone = ({ projectName, repoPath }) =>
+  new Promise((resolve, reject) => {
```
This outer Promise is also unneccessary. This could be refactored as:

```js
const clone = ({ projectName, repoPath }) =>
  git()
    .silent(true)
    .clone(repoPath, projectName)
```

jasonkylefrank replied:
@stephenmathieson 
Here's some further info.  The "long and the short of it" though is that I think it will actually be better if we keep our `Promise` wrappers in our `clone()` and `validateRepoName()` functions.

The reason I say that is because the `simple-git` library is returns a reference to its main `Git` object from each of its public methods.  That allows its chaining.  Furthermore, [they implemented a `then()` method](https://github.com/steveukx/git-js/blob/ef368c87fb4e1be5bbd40b1644cc8475e7bd6915/src/git.js#L1192) (but they have since deprecated it).  Although that implementation will run the function passed to it, it still returns a reference its main `Git` object.  So I think that is why I was getting that runtime error on that line that had a `.catch()` at the end of the `.then()`'s  - because it was trying to call `.catch()` on the **`Git`** object.  

So I believe we actually want our own `Promise` wrappers to keep the code consistent with what you've done elsewhere, with your `.then()` and `.catch()` series.

Thoughts? 😀

stephenmathieson commented on lib/utils/clone-repo.js:
```diff
@@ -0,0 +1,52 @@
+    git().silent(true)
+      .clone(repoPath, projectName)
+      .then(resolve)
+      // TODO: Figure out why including this catch block screws things up
```
What is meant by "screws things up" here?

jasonkylefrank replied:
@stephenmathieson When that `catch()` block was there two things happened that blocked this implementation from working.  Even when I supplied a valid repo path argument it would:

1.  The promise chain would be in the "rejected state" (causing the `catch` block on line 47 to run).  That is to say, this catch block on line 27 would execute, even when it seemed like it should not.  That said, see the next issue...
2. Even though it seemed that this `catch` block executed every time, code that I put in that block did not seem to execute.  For example, a `console.log()` would not do anything.

I'm sure you can enlighten me as to why these two issues were occurring!  😀

stephenmathieson commented on package.json:
```diff
@@ -31,8 +31,12 @@
     "make-dir": "^1.3.0",
     "ms": "^1.0.0",
     "mz": "^2.6.0",
+    "next": "7.0.1",
```
Why was this dependency added?

jasonkylefrank replied:
@stephenmathieson 
I removed that dependency.  Once I fix the other items I'll update this PR with all of the changes.

stephenmathieson commented:
This is great Jason, thanks! I've left a few comments inline we should discuss, but I think we can get this in soon 😄

jasonkylefrank commented:
@stephenmathieson 
Thanks for your comments!  I've tried to address each one.  I'm currently blocked on [this comment thread](https://github.com/segmentio/create-next-app/pull/48#discussion_r223020831) though.

franky47 commented:
I'd be very interested in this as well.

As for the argument naming, I instinctively typed:
```
yarn create react-app my-project -e https://github.com/franky47/next-template
```
Although -e should probably be reserved for "official" examples in the nextjs repository.

bbuckland commented:
Would be interested in using this @jasonkylefrank — currently leveraging create react app fork/template solution, but looking to create something for next. Where are you at with this? Can I help?

jasonkylefrank commented:
Hi @bbuckland !  Great to hear that you're interested in this!

To answer your question about the status, I think this PR is very close to being accepted.  I was only waiting on response from @stephenmathieson regarding my Oct 12 fixes.  

I believe I addressed the issues that @stephenmathieson pointed out, but 2 small issues require some interaction to finalize:

1. The command-line keyword `--clone-repo` could be adjusted if we find that another keyword is better.  In [this comment thread](https://github.com/segmentio/create-next-app/pull/48/files#r224909459) I suggested some options.

2. There was some discussion about whether or not we needed some `Promise` wrappers in some of the new code, particularly in my `validateRepoName()` and `clone()` functions in `clone-repo.js`.  After digging into it more, I think we should keep them present.  This is due to how the `simple-git` library uses promises internally.  I outlined this issue more in [this comment thread](https://github.com/segmentio/create-next-app/pull/48#discussion_r224907510). 

So in short, if we (a) agree on the proper keyword and (b) are agreeable to use the promise wrappers mentioned in `clone-repo.js`, then I believe this is done.  Again, mostly just waiting on response from @stephenmathieson .

Final note, it looks like my code in `clone-repo.js` currently has the promise wrapper in `clone()` and does not have the promise wrapper in `validateRepoName()`.  That's because I was systematically trying to get rid of those wrappers when I ran into those problems and ended up finding the info about how the `simple-git` library is using promises internally.  I think my testing showed that it would be better to keep the promise wrapper in `validateRepoName()` (just like it currently is in `clone()`.  I can't test that at the moment though, so maybe that's something that @bbuckland can help test (with and without the promise wrapper in `validateRepoName()`)?
